nrf_modem: v2.0.0 (#734)
nrfconnect/sdk-nrfxlib: nrf_modem/doc/fault_handling.rst

.. __fault_handling:

Modem fault handling
####################

> The application core and the modem core are separate cores on nRF9160.

Reviewer: ```suggestion
The application core and the modem core are separate cores on an nRF9160 SiP.
```

> When the modem core on the nRF9160 crashes, it sends a fault signal to the application core.

Reviewer: ```suggestion
When the modem core on the nRF9160 SiP crashes, it sends a fault signal to the application core.
```

Detecting a fault
*****************

> The Modem library implements a fault handling mechanism allowing the application to be notified when the modem core has crashed.

Reviewer: ```suggestion
The Modem library implements a fault handling mechanism, allowing the application to be notified when the modem core has crashed.
```

> The application can set a modem fault handler function during initialization using the Modem library initialization parameters.

Reviewer: ```suggestion
The application can set a modem fault handler function using the Modem library initialization parameters during initialization.
```

> For a complete list of modem fault reasons, see :ref:`nrf_modem_fault`.

Reviewer: Where is this linking to? Was it supposed to link to a section in API?

Author: it's in `nrf_modem.h`

Reviewer: Oh, I am not sure if we can directly link to it, unless its pulled into the API.rst file.

Author: Ok, I'll add it to `api.rst`

> In |NCS| the :ref:`nrf_modem_lib_readme` sets the modem fault handler during Modem library initialization.

Reviewer: ```suggestion
In |NCS|, the :ref:`nrf_modem_lib_readme` sets the modem fault handler during Modem library initialization.
```

> How to recover from a fault
> ***************************

Reviewer: ```suggestion
Recovering from a fault
***********************
```

> When the modem has crashed, it can be re-initialized by the application.
> The application core does not need to be reset in order to recover the modem from a fault state.
> The application can re-initialize the modem by re-initializing the Modem library, by calling :c:func:`nrf_modem_shutdown` and :c:func:`nrf_modem_init`.

Reviewer: ```suggestion
When the modem has crashed, it can be reinitialized by the application.
The application core does not need to be reset to recover the modem from a fault state.
The application can reinitialize the modem by reinitializing the Modem library through calls to :c:func:`nrf_modem_shutdown` and :c:func:`nrf_modem_init` functions.
```

> .. important:
>    The re-initialization of the Modem library must be done outside of the Modem library fault handler.

Reviewer: ```suggestion
.. important::

   The reinitialization of the Modem library must be done outside of the Modem library fault handler.
```

Networking sockets
******************

> Socket APIs which require communication with the modem, return -1 and set ``errno`` to ``NRF_ESHUTDOWN`` if the modem has crashed or if it is uninitialized.

Reviewer: ```suggestion
Socket APIs that require communication with the modem, return ``-1`` and set ``errno`` to ``NRF_ESHUTDOWN`` if the modem has crashed or if it is uninitialized.
```

> Although the modem has crashed, any data which was stored by the Modem library, including data that was delivered to Modem library by the modem, will remain available until the Modem library is shut down.

Reviewer: ```suggestion
Although the modem has crashed, any data which was stored by the Modem library, including data that was delivered to Modem library by the modem, remains available until the Modem library is shut down.
```

> This data includes incoming network data which was received before the crash but not yet read by the application.

Reviewer: ```suggestion
This data includes incoming network data, which was received before the crash but has not been read by the application.
```

The application can read that data as normal, using the :c:func:`recv` function.
When no more data is available, the :c:func:`recv` returns -1 and sets ``errno`` to ``NRF_ESHUTDOWN``.

Ongoing API calls
*****************

It may happen that the modem crashes when an function call in Modem library is ongoing, for example, a socket API call.
When the Modem library detects that a fault has occurred in the modem, it will immediately exit all calls which are waiting for a modem response.

We can distinguish between two categories of APIs in Modem library:

> # Socket APIs, exit immediately returning -1 and setting errno to ``NRF_ESHUTDOWN``
> # Non-Socket APIs, exit immediately returning ``-NRF_ESHUTDOWN``

Reviewer: ```suggestion
* Socket APIs - Exit immediately returning ``-1`` and sets errno to ``NRF_ESHUTDOWN``
* Non-Socket APIs - Exit immediately returning ``-NRF_ESHUTDOWN``
```

> .. __fault_handling:

Reviewer: ```suggestion
.. _fault_handling:
```
